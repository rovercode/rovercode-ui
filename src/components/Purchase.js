import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import {
  Button,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  CardMedia,
  Chip,
  CircularProgress,
  Container,
  Grid,
  Typography,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { Alert } from '@material-ui/lab';
import { loadStripe } from '@stripe/stripe-js';
import { FormattedMessage, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

import gigglebotImage from '@/assets/images/connection-help/gigglebot.png';

const styles = (theme) => ({
  media: {
    height: 0,
    paddingTop: '70%',
  },
  mainContainer: {
    marginBottom: theme.spacing(8),
    [theme.breakpoints.up('xs')]: {
      minWidth: theme.breakpoints.values.xs,
    },
    [theme.breakpoints.up('sm')]: {
      minWidth: theme.breakpoints.values.sm,
    },
    [theme.breakpoints.up('md')]: {
      minWidth: theme.breakpoints.values.md,
    },
    [theme.breakpoints.up('lg')]: {
      minWidth: theme.breakpoints.values.lg,
    },
  },
  card: {
    marginTop: theme.spacing(4),
  },
  settingsSection: {
    marginBottom: theme.spacing(4),
  },
  title: {
    fontWeight: 'bold',
  },
  titleArea: {
    marginBottom: theme.spacing(4),
  },
});

class Purchase extends Component {
  componentDidMount() {
    const { fetchSubscription, user } = this.props;
    fetchSubscription(user.user_id);
  }

  componentDidUpdate(prevProps) {
    const { isCreating: wasCreatingCheckout } = prevProps;
    const { isCreating: isCreatingCheckout } = this.props;
    if (wasCreatingCheckout && !isCreatingCheckout) {
      return this.redirectToStripe();
    }
    return null;
  }

  redirectToStripe = () => {
    const { checkoutSessionId } = this.props;
    return loadStripe(STRIPE_SHARABLE_KEY) // eslint-disable-line no-undef
      .then((stripe) => {
        stripe
          .redirectToCheckout({
            sessionId: checkoutSessionId,
          });
      });
  }

  handleCheckoutClick = () => {
    const { createCheckoutSession, user } = this.props;
    createCheckoutSession(
      user.user_id.toString(),
      [{ name: 'Individual Plan', quantity: 1 }], // In the future, this is user selectable
      `${window.location.origin}/user/settings`, // Success
      `${window.location.origin}/purchase`, // Failure
      true,
    );
  }

  render() {
    const {
      classes,
      isFetching,
      subscription,
      creationError,
      intl,
    } = this.props;

    const costText = intl.formatMessage({
      id: 'app.friends_family.cost',
      description: 'Price label',
      defaultMessage: '$150.00 billed annually',
    });

    const productTitle = intl.formatMessage({
      id: 'app.friends_family.product_title',
      description: 'Individual plan product title',
      defaultMessage: 'Individual Subscription with Rover',
    });
    

    const checkoutSessionErrorMessage = () => (
      <Grid item>
        <Alert variant="outlined" severity="error">
          <FormattedMessage
            id="app.friends_family.error_message"
            description="Message for an error creating a checkout session"
            defaultMessage="There was an error. Please contact support@rovercode.com to complete your purchase."
          />
        </Alert>
      </Grid>
    );

    const pageContent = () => {
      if (isFetching) {
        return (
          <Grid item xs={12} lg={12}>
            <CircularProgress />
          </Grid>
        );
      }
      if (subscription.plan === '3') {
        return (
          <>
            <Grid item xs={12} lg={12}>
              <Typography>
                <FormattedMessage
                  id="app.friends_family.already_purchased"
                  description="Messages explaining that the user already purchased the Individual Plan"
                  defaultMessage="You have already purchased the Individual Plan for this account"
                />
              </Typography>
            </Grid>
            <Button
              variant="contained"
              color="primary"
              component={Link}
              to="/courses"
            >
              <Typography>
                <FormattedMessage
                  id="app.friends_family.already_purchased_action"
                  description="Button giving an action for those who have already purchased"
                  defaultMessage="Start your first lesson!"
                />
              </Typography>
            </Button>
          </>
        );
      }
      return (
        <>
          <Grid item xs={12} lg={12}>
            <Typography variant="subtitle1">
              <FormattedMessage
                id="app.friends_family.instructions"
                description="Paragraph explaining the friends-and-family purchase page"
                defaultMessage="We're excited to offer you early access to Rovercode!
                Shop below, and please email support@rovercode.com with any questions or issues."
              />
            </Typography>
          </Grid>
          <Grid item xs={12} lg={4}>
            <Card className={classes.card}>
              <CardHeader title={productTitle} />
              <CardMedia
                className={classes.media}
                image={gigglebotImage}
                title="Gigglebot Rover"
              />
              <CardContent>
                <Typography variant="body2">
                  <FormattedMessage
                    id="app.friends_family.item_description"
                    description="Description of the individual plan"
                    defaultMessage="One year of unlimited access to the Rovercode platform and curriculum.
                    Your first year includes a Gigglebot Rover. If you have a coupon code, you can enter it on the next screen.
                    Your subscription will automatically renew in one year."
                  />
                </Typography>
              </CardContent>
              <CardActions>
                <Button
                  variant="contained"
                  color="primary"
                  onClick={this.handleCheckoutClick}
                >
                  <FormattedMessage
                    id="app.purchase.checkout"
                    description="Button label to checkout purchase"
                    defaultMessage="Check Out"
                  />
                </Button>
                <Chip label={costText} />
              </CardActions>
            </Card>
          </Grid>
        </>
      );
    };

    return (
      <Container className={classes.mainContainer}>
        <Grid className={classes.settingsSection} container spacing={2}>
          <Grid className={classes.titleArea} item container direction="row" justify="space-between">
            <Grid item>
              <Typography className={classes.title} variant="h4">
                Friends and Family Early Access
              </Typography>
            </Grid>
          </Grid>
        </Grid>
        <Grid className={classes.settingsSection} container spacing={2}>
          <Grid item container direction="row" alignItems="flex-start" spacing={1}>
            {creationError ? checkoutSessionErrorMessage() : <></> }
            {pageContent()}
          </Grid>
        </Grid>
      </Container>
    );
  }
}

Purchase.defaultProps = {
  checkoutSessionId: null,
  subscription: {
    plan: '',
  },
  creationError: null,
};

Purchase.propTypes = {
  checkoutSessionId: PropTypes.string,
  classes: PropTypes.shape({
    card: PropTypes.string.isRequired,
    mainContainer: PropTypes.string.isRequired,
    media: PropTypes.string.isRequired,
    settingsSection: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    titleArea: PropTypes.string.isRequired,
  }).isRequired,
  createCheckoutSession: PropTypes.func.isRequired,
  fetchSubscription: PropTypes.func.isRequired,
  user: PropTypes.shape({
    user_id: PropTypes.number.isRequired,
  }).isRequired,
  subscription: PropTypes.shape({
    plan: PropTypes.string,
  }),
  isFetching: PropTypes.bool.isRequired,
  isCreating: PropTypes.bool.isRequired,
  creationError: PropTypes.string,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default injectIntl(withStyles(styles)(Purchase));
