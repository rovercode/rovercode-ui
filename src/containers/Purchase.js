import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { checkAuthError, authHeader } from '@/actions/auth';
import { fetchSubscription } from '@/actions/subscription';
import { createCheckoutSession } from '@/actions/checkout';
import Purchase from '@/components/Purchase';

const mapStateToProps = ({ checkout, user, subscription }) => (
  { ...checkout, user, ...subscription }
);
const mapDispatchToProps = (dispatch, { cookies }) => ({
  createCheckoutSession: (
    userId, lineItems, successURL, cancelURL, collectShippingAddress,
  ) => dispatch(createCheckoutSession(
    userId, lineItems, successURL, cancelURL, collectShippingAddress, authHeader(cookies),
  )).catch(checkAuthError(dispatch)),
  fetchSubscription: (id) => dispatch(fetchSubscription(id, authHeader(cookies))),
});

const PurchaseContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(Purchase);

PurchaseContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(PurchaseContainer));
