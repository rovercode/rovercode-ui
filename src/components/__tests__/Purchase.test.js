import React from 'react';
import { loadStripe } from '@stripe/stripe-js';
import { Button } from '@material-ui/core';
import Purchase from '../Purchase';

jest.mock('@stripe/stripe-js');

let fetchSubscription;
let createCheckoutSession;

describe('The Purchase component', () => {
  beforeEach(() => {
    loadStripe.mockClear();
    fetchSubscription = jest.fn().mockResolvedValue();
    createCheckoutSession = jest.fn().mockResolvedValue();
  });

  test('renders on the page with no errors', () => {
    const user = {
      user_id: 1,
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
      />,
    ).dive().dive().dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('shows error message on checkout session creation error', () => {
    const user = {
      user_id: 1,
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
        creationError="Everything went super wrong."
      />,
    ).dive().dive().dive();
    expect(wrapper.find('WithStyles(ForwardRef(Alert))').exists()).toBe(true);
  });

  test('shows circular progress while fetching subscription', () => {
    const user = {
      user_id: 1,
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
        isFetching
      />,
    ).dive().dive().dive();
    expect(wrapper.find('WithStyles(ForwardRef(CircularProgress))').exists()).toBe(true);
  });

  test('shows already purchased message when already purchased', () => {
    const user = {
      user_id: 1,
    };
    const subscription = {
      plan: '3',
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
        subscription={subscription}
      />,
    ).dive().dive().dive();
    expect(wrapper.find('FormattedMessage').at(0)
      .prop('defaultMessage'))
      .toEqual(expect.stringContaining('You have already purchased'));
  });

  test('creates a checkout session on button click', () => {
    const user = {
      user_id: 1,
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
      />,
    ).dive().dive().dive();
    wrapper.find(Button).simulate('click');
    expect(createCheckoutSession).toHaveBeenCalled();
  });

  test('redirects to Stripe once the checkout session is created', (done) => {
    const redirectToCheckout = jest.fn();
    loadStripe.mockResolvedValue({ redirectToCheckout });
    const user = {
      user_id: 1,
    };
    const wrapper = shallowWithIntl(
      <Purchase
        user={user}
        fetchSubscription={fetchSubscription}
        createCheckoutSession={createCheckoutSession}
      />,
    ).dive().dive();
    wrapper.setProps({
      isCreating: false,
    });
    const prevProps = {
      isCreating: true,
    };
    wrapper.dive().instance().componentDidUpdate(prevProps).then(() => {
      expect(redirectToCheckout).toHaveBeenCalled();
      done();
    });
  });
});
