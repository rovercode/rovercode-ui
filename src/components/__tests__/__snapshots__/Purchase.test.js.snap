// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Purchase component renders on the page with no errors 1`] = `
<WithStyles(ForwardRef(Container))
  className="Purchase-mainContainer-2"
>
  <WithStyles(ForwardRef(Grid))
    className="Purchase-settingsSection-4"
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      className="Purchase-titleArea-6"
      container={true}
      direction="row"
      item={true}
      justify="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Typography))
          className="Purchase-title-5"
          variant="h4"
        >
          Friends and Family Early Access
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    className="Purchase-settingsSection-4"
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      alignItems="flex-start"
      container={true}
      direction="row"
      item={true}
      spacing={1}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        lg={12}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          variant="subtitle1"
        >
          <FormattedMessage
            defaultMessage="We're excited to offer you early access to Rovercode! Shop below, and please email"
            id="app.friends_family.instructions_1"
          />
           
          <WithStyles(ForwardRef(Link))
            href="mailto:support@rovercode.com"
          >
            support@rovercode.com
          </WithStyles(ForwardRef(Link))>
           
          <FormattedMessage
            defaultMessage="with any questions or issues."
            id="app.friends_family.instructions_2"
          />
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        lg={4}
        xs={12}
      >
        <WithStyles(ForwardRef(Card))
          className="Purchase-card-3"
        >
          <WithStyles(ForwardRef(CardHeader))
            title="Individual Subscription with Rover"
          />
          <WithStyles(ForwardRef(CardMedia))
            className="Purchase-media-1"
            image={Object {}}
            title="Gigglebot Rover"
          />
          <WithStyles(ForwardRef(CardContent))>
            <WithStyles(ForwardRef(Typography))
              variant="body2"
            >
              <FormattedMessage
                defaultMessage="One year of unlimited access to the Rovercode platform and curriculum. Your first year includes a Gigglebot Rover. If you have a coupon code, you can enter it on the next screen. Your subscription will automatically renew in one year."
                id="app.friends_family.item_description"
              />
            </WithStyles(ForwardRef(Typography))>
          </WithStyles(ForwardRef(CardContent))>
          <WithStyles(ForwardRef(CardActions))>
            <WithStyles(ForwardRef(Button))
              color="primary"
              onClick={[Function]}
              variant="contained"
            >
              <FormattedMessage
                defaultMessage="Check Out"
                id="app.purchase.checkout"
              />
            </WithStyles(ForwardRef(Button))>
            <WithStyles(ForwardRef(Chip))
              label="$160.00 billed annually"
            />
          </WithStyles(ForwardRef(CardActions))>
        </WithStyles(ForwardRef(Card))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
